hashintel/hash
# H-2270: Create active research workflow for inferring entities from web search results

## 🌟 What is the purpose of this PR?




This PR adds a `startResearchTask` temporal workflow, which obtains entities of a given type by:
1. performing a web-search query to an external endpoint
2. proposing entities based on the textual content of the web-pages (in parallel)
3. creating draft entities for each proposed

A significant portion of the diff is re-using parts of the existing `inferEntities` temporal activity, and exposing the functionality as individual activities.


## 🔍 What does this change?




- adds an internal API client generator to the yarn workspace, to allow for generated return-types from the API client

## 🔗 Related links





- H-2270

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



- allow for setting limi

## 🐾 Next steps



- allow for indicating how many entities the research task should create (currently this isn't specified)
- ask an LLM to create one or more specific web search queries, rather than re-using the prompt as the web search query
- deduplicate the proposed entities returned from different webpages before trying to create them
- persist more information in the graph regarding the completed flow (e.g. the prompt use, the time it took to complete the task, the web pages used to obtain the result, etc)

## 🛡 What tests cover this?



None

## ❓ How to test this?



1. Follow the steps in https://github.com/hashintel/internal-api to run the internal API locally at `localhost:5002`
2. Rebuild the external services, or run the `hash-ai-worker-ts` locally
3. Login to the FE
4. Create a type which you want to research entities for (e.g. `Person`, `Company`)
5. Navigate to the new AI tab in the sidebar, and enter a prompt and select the type you created
6. Click "Start Research Task"
7. Observe the created draft entities (or references to existing entities if they came up in the search results)

## 📹 Demo



<img width="1286" alt="image" src="https://github.com/hashintel/hash/assets/42802102/25a55d7e-d904-4aa6-80a9-bb03091eabcf">
